kore/test/Test/Kore/Unification/Unifier.hs

```diff
@@ -646,81 +560,119 @@ injUnificationTests :: [TestTree]
+            (Mock.sortInjectionSubToTop
+                (Mock.sortInjectionSubSubToSub (mkVar Mock.xSubSubSort))
+            )
         andSimplifySuccess
             term1
-            (UnificationTerm (inj s2 a))
+            (UnificationTerm (Mock.sortInjectionSubSubToTop Mock.aSubSubsort))
             [ UnificationResult
-                { term = inj s2 a
-                , substitution = [("x", a)]
+                { term = Mock.sortInjectionSubSubToTop Mock.aSubSubsort
+                , substitution = [("xSubSubSort", Mock.aSubSubsort)]
                 , predicate = Syntax.Predicate.makeTruePredicate
                 }
             ]
     , testCase "doubly injected variable vs doubly injected term" $ do
-        term1 <- simplifyPattern $ UnificationTerm (inj s2 (inj s4 x))
-        term2 <- simplifyPattern $ UnificationTerm (inj s2 (inj s3 a))
+        term1 <- simplifyPattern $ UnificationTerm
+            (Mock.sortInjectionSubToTop
+                (Mock.sortInjectionSubSubToSub Mock.aSubSubsort)
+            )
+        term2 <- simplifyPattern $ UnificationTerm
+            (Mock.sortInjectionOtherToTop
+                (Mock.sortInjectionSubSubToOther (mkVar Mock.xSubSubSort))
+            )
         andSimplifySuccess
             term1
             term2
             [ UnificationResult
-                { term = inj s2 a
-                , substitution = [("x", a)]
+                { term = Mock.sortInjectionSubSubToTop Mock.aSubSubsort
+                , substitution = [("xSubSubSort", Mock.aSubSubsort)]
                 , predicate = Syntax.Predicate.makeTruePredicate
                 }
             ]
     , testCase "constant vs injection is bottom" $
         andSimplifySuccess
-            (UnificationTerm a)
-            (UnificationTerm (inj s1 xs2))
+            (UnificationTerm Mock.aTopSort)
+            (UnificationTerm (Mock.sortInjectionSubSubToTop Mock.aSubSubsort))
             []
     , testCase "unmatching nested injections" $ do
         term1 <-
             simplifyPattern
-            $ UnificationTerm (inj s4 (inj s2 a))
+            $ UnificationTerm
+                (Mock.sortInjectionSubToTop
+                    (Mock.sortInjectionSubSubToSub Mock.aSubSubsort)
+                )
         term2 <-
             simplifyPattern
-            $ UnificationTerm (inj s4 (inj s3 b))
+            $ UnificationTerm
+                (Mock.sortInjectionOtherToTop
+                    (Mock.sortInjectionSubOtherToOther Mock.aSubOthersort)
+                )
         andSimplifySuccess
             term1
             term2
             []
+    , testCase "matching injections" $
+        andSimplifySuccess
+            (UnificationTerm (Mock.sortInjectionSubSubToTop Mock.aSubSubsort))
+            (UnificationTerm (Mock.sortInjectionSubToTop (mkVar Mock.xSubSort)))
+            [ UnificationResult
+                { term =
+                    -- TODO (virgil): Fix this in unification, injection
```

Comment:
Not relevant to this pull request directly, but: In light of #681, we see that injections are actually regarded as constructors in the sorts where they appear. Couldn't we normalize sort injections by fully expanding them, i.e. so that in `inj{from, to}` the sorts `from` and `to` are always in a direct subsort relation?

Reply:
It depends. Are we going to handle diamond-like subsort relations (a is a subsort of b and c, which are subsorts of c)? There was a long discussion about it a while ago, and I don't remember the outcome. My guess is that we want to handle those, at least to some extent, and, if so, I think unification would become more complicated if we were using only direct subsort relations and we'd be forced to pick one of the two branches of the diamond.

Reply:
b and c are subsorts of d above, sorry.

Reply:
If I understand the result of #681, then diamond relations should be invalid, but I will bring it up there so we get an explicit answer.